spark3/src/main/java/org/apache/iceberg/actions/Spark3SnapshotAction.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.actions;
+
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.SnapshotSummary;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.TableProperties;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.spark.SparkTableUtil;
+import org.apache.iceberg.spark.source.SparkTable;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.catalyst.TableIdentifier;
+import org.apache.spark.sql.catalyst.analysis.NoSuchNamespaceException;
+import org.apache.spark.sql.catalyst.analysis.TableAlreadyExistsException;
+import org.apache.spark.sql.connector.catalog.CatalogPlugin;
+import org.apache.spark.sql.connector.catalog.Identifier;
+import org.apache.spark.sql.connector.catalog.StagedTable;
+import org.apache.spark.sql.connector.catalog.TableCatalog;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.collection.JavaConverters;
+
+/**
+ * Creates a new Iceberg table based on a source Spark table. The new Iceberg table will
+ * have a different data and metadata directory allowing it to exist independently of the
+ * source table.
+ */
+class Spark3SnapshotAction extends Spark3CreateAction implements SnapshotAction {
+  private static final Logger LOG = LoggerFactory.getLogger(Spark3SnapshotAction.class);
+
+  private String destTableLocation = null;
+
+  Spark3SnapshotAction(SparkSession spark, CatalogPlugin sourceCatalog,
+                       Identifier sourceTableName, CatalogPlugin destCatalog,
+                       Identifier destTableName) {
+    super(spark, sourceCatalog, sourceTableName, destCatalog, destTableName);
+  }
+
+  @Override
+  public Long execute() {
+    StagedTable stagedTable;
+    Table icebergTable;
+
+    try {
+      stagedTable = destCatalog().stageCreate(destTableName(), v1SourceTable().schema(),
+          sourcePartitionSpec(), buildPropertyMap());
+      icebergTable = ((SparkTable) stagedTable).table();
```

Comment:
How about we add a check that the table location, as well as metadata and data locations, do not interfere with the original table? This way we don't have to know the default table location the catalog will assign.

Reply:
We can just add a TODO now and address it later as the probability of hitting this is low.